File: py/math/EulerTotientFunction.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EulerTotientFunction:
    def solution(self, num: int) -> List[int]:
        isPrime, primes = [True for _ in range(num)], []
        result = [0 for _ in range(num)]
        for i in range(2, num + 1):
            if isPrime[i - 1]:
                primes.append(i)
                result[i - 1] = i - 1

            for value in primes:
                if value > num // i:
                    break

                isPrime[i * value - 1] = False
                if not (i % value):
                    result[i * value - 1] = result[i - 1] * value
                    break
                else:
                    result[i * value - 1] = result[i - 1] * (value - 1)
        return result

File: java/math/EulerTotientFunction.java
package math;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class EulerTotientFunction {
    public int[] solution(int num) {
        boolean[] isPrime = new boolean[num];
        Arrays.fill(isPrime, true);
        List<Integer> primes = new ArrayList<>();

        int[] result = new int[num];
        Arrays.fill(result, 0);
        for (int i = 2; i <= num; ++i) {
            if (isPrime[i - 1]) {
                primes.add(i);
                result[i - 1] = i - 1;
            }

            for (int value : primes) {
                if (value > num / i) {
                    break;
                }

                isPrime[i * value - 1] = false;
                if (0 == i % value) {
                    result[i * value - 1] = result[i - 1] * value;
                    break;
                } else {
                    result[i * value - 1] = result[i - 1] * (value - 1);
                }
            }
        }
        return result;
    }
}
